### PR TITLE
Communication R4 Read, Search, Write, and Patch

### PR DESCRIPTION
**Pull request by sydneanschutz**
Description
----
Communication R4 Read, Search, Write, and Patch
Screen Shot 2021-04-06 at 5 00 33 PM
Screen Shot 2021-04-06 at 5 00 36 PM
Screen Shot 2021-04-06 at 5 01 48 PM

PR Checklist
----
- [x] Screenshot(s) of changes attached before changes merged.
- [ ] Screenshot(s) of changes attached after changes merged and published.

### CODE REVIEW

File: content/millennium/r4/other/binary.md [hunk 1]
```diff
@@ -9,7 +9,7 @@ title: Binary | R4 API
 
 ## Overview
 
-The Binary resource can contain any clinical content such as text, images, and PDFs. This resource is currently limited to Continuity of Care Documents (CCD), clinical documents and diagnostic reports. The diagnostic reports are currently limited to Anatomic Pathology and Radiology.
+The Binary resource can contain any clinical content such as text, images, and PDFs. This resource is currently limited to clinical documents, diagnostic reports, and communication message content. Diagnostic reports currently supports Anatomic Pathology, Cardiology, Microbiology, and Radiology.
```
Comment (sydneanschutz) [inline content/millennium/r4/other/binary.md:12]: Wasn't sure on the wording here after the rebase

Comment (kathypickering) [reply content/millennium/r4/other/binary.md:12]: I think we should keep Continuity of Care Documents (CCD)

Comment (ebaranowski) [reply content/millennium/r4/other/binary.md:12]: I would leave CCD in the list, just easier to see buzz words in the list.

Also: 'The supported diagnostic reports are...'

File: content/millennium/r4/other/binary.md [hunk 2]
```diff
@@ -42,7 +42,7 @@ _Implementation Notes_
 
 _Implementation Notes_
 
-* The `Binary.read` scope and either the `DiagnosticReport.read` or the `DocumentReference.read` scope is required.
+* The `Binary.read` scope and either one of the `DiagnosticReport.read`, the `DocumentReference.read`, or the `Communication.read` scope is required.
```
Comment (kathypickering) [inline content/millennium/r4/other/binary.md:45]: Can I retrieve a DiagnosticReport or DocumentReference with the Communication scope?  No, right?
I think you can flip Communication first 
The `Binary.read` scope and `Communication.read` scope or either the `DiagnosticReport.read` or the `DocumentReference.read` is required.

File: content/millennium/r4/request-and-response/communication.md [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+ `_id`             | This or (`category`, `-email-status`, `recipient` and `received`) | [`token`]     | The logical resource id associated with the resource. Example: `489580643.0.-4.prsnl`
+ `category`        | (This, `recipient`, `received` and `-email-status`) or `_id`      | [`token`]     | A token for a [`CodeableConcept`] that points to the CodeableConcept used for Communication.category. Example: `http://terminology.hl7.org/CodeSystem/communication-category|notification`
+ `recipient`       | (This, `category`, `received` and `-email-status`) or `_id`       | [`reference`] | The recipient of the communication. Example: `Practitioner/3456783`
+ `received`        | (This, `category`, `recipient` and `-email-status`) or `_id`      | [`date`]      | Date range into which the communication falls. Example: `le2017-02-01T10:30:00Z`
+ `-email-status`   | (This, `category`, `recipient` and `received`) or `_id`           | [`token`]     | The status of the email. Example: `http://hl7.org/fhir/task-status|in-progress`
+
+
+ Notes:
+
+ * The `recipient` parameter must represent a Practitioner or a Group. Use the `:Group` modifier to search by Group.
```
Comment (ebaranowski) [inline content/millennium/r4/request-and-response/communication.md:72]: Remove this:

'Use the `:Group` modifier to search by Group.'

File: content/millennium/r4/request-and-response/communication.md [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+ * The `received` parameter may be provided up to two times, and must use the `ge` or `le` prefixes. When provided twice, the lower value must have a `ge` prefix and the higher value must have a `le` prefix. The `ge` prefix is only supported on one.
```
Comment (kathypickering) [inline content/millennium/r4/request-and-response/communication.md:73]: The 'ge' prefix is only supported when provided once??

**Review by ebaranowski — CHANGES_REQUESTED** [review]

File: content/millennium/r4/other/binary.md (hunk 2 above)
Comment (ebaranowski) [reply content/millennium/r4/other/binary.md:45]: I mean.... This is just a giant or.... which is a list like it's already written. I think you want something different like a list of options (which might not be able to be conveyed in a super long list)

```
Communication.read scope or either the DiagnosticReport.read or

File: content/millennium/r4/request-and-response/communication.md (hunk 4 above)
Comment (ebaranowski) [reply content/millennium/r4/request-and-response/communication.md:73]: ge prefix is only supported on one of the received parameters. We could probably just say it's required

File: content/millennium/r4/request-and-response/communication.md [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+The following fields are returned if valued:
+
+* [Communication Id](https://hl7.org/FHIR/communication-definitions.html#Communication.identifier)
+* [Status (Completed)](https://hl7.org/FHIR/communication-definitions.html#Communication.status)
+* [Category](https://hl7.org/FHIR/communication-definitions.html#Communication.category)
+* [Priority](https://hl7.org/FHIR/communication-definitions.html#Communication.priority)
+* [Subject](https://hl7.org/FHIR/communication-definitions.html#Communication.subject)
+* [Topic](https://hl7.org/FHIR/communication-definitions.html#Communication.topic)
+* [Encounter](https://hl7.org/FHIR/communication-definitions.html#Communication.encounter)
+* [Sent DateTime](https://hl7.org/FHIR/communication-definitions.html#Communication.sent)
```
Comment (VelocityKnight) [inline content/millennium/r4/request-and-response/communication.md:23]: The 'In Response To' field needs to be added. https://hl7.org/FHIR/communication-definitions.html#Communication.inResponseTo

File: lib/resources/example_json/r4_examples_communication.rb [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+                'code': 'completed',
+                'display': 'Completed'
+              }
+            ],
+            'text': 'Completed'
+          },
+          'url': 'https://fhir-ehr.cerner.com/r4/StructureDefinition/email-status'
+        }
+      ],
+      'status': 'completed',
```
Comment (VelocityKnight) [inline lib/resources/example_json/r4_examples_communication.rb:32]: Add inResponseTo between lines 31 and 32, e.g.
``` 
               "inResponseTo": [
                    {
                        "reference": "Communication/512657279.0.-4.prsnl"
                    }
                ]
```

File: content/millennium/r4/other/binary.md (hunk 1 above)
Comment (sydneanschutz) [reply content/millennium/r4/other/binary.md]: Added back CCD and I checked with Mayur; cardio and micro should be included on this list.

https://github.com/cerner/fhir.cerner.com/pull/607/commits/99c1be98f1d5fe1d1fa6c1cfa9b11d4c798094b9#diff-b9330978e62a15a61d6612ae8dc779e3b4680761c2dfc7e9b3778909d2d95cd8R12

File: content/millennium/r4/request-and-response/communication.md (hunk 5 above)
Comment (sydneanschutz) [reply content/millennium/r4/request-and-response/communication.md]: https://github.com/cerner/fhir.cerner.com/pull/607/commits/99c1be98f1d5fe1d1fa6c1cfa9b11d4c798094b9#diff-cc6586eb639048bdea01208ce786612c64361622cd67bd53a5bf7253f27a840aR17

File: content/millennium/r4/request-and-response/communication.md (hunk 3 above)
Comment (sydneanschutz) [reply content/millennium/r4/request-and-response/communication.md]: https://github.com/cerner/fhir.cerner.com/pull/607/commits/99c1be98f1d5fe1d1fa6c1cfa9b11d4c798094b9#diff-cc6586eb639048bdea01208ce786612c64361622cd67bd53a5bf7253f27a840aL72

File: lib/resources/example_json/r4_examples_communication.rb (hunk 6 above)
Comment (sydneanschutz) [reply lib/resources/example_json/r4_examples_communication.rb:37]: https://github.com/cerner/fhir.cerner.com/pull/607/commits/99c1be98f1d5fe1d1fa6c1cfa9b11d4c798094b9#diff-7b749c1dd8b022a87bedf2a50fba0c2ffa5fa0b910c5d15a1cb40011ba9004deR32

File: content/millennium/r4/other/binary.md (hunk 2 above)
Comment (kathypickering) [reply content/millennium/r4/other/binary.md:45]: I think Evan and I are saying the same thing.  Evan just didn't finish the sentence.

File: content/millennium/r4/request-and-response/communication.md (hunk 4 above)
Comment (kathypickering) [reply content/millennium/r4/request-and-response/communication.md:74]: Can the received simply be a date?  If so, I wouldn't say it is required.
'When provided once, only the 'ge' prefix is supported'

Comment (ebaranowski) [reply content/millennium/r4/request-and-response/communication.md:74]: if it's just a date then it defaults to eq and that is not supported. So no

File: content/millennium/r4/other/binary.md (hunk 2 above)
Comment (ebaranowski) [reply content/millennium/r4/other/binary.md:45]: I did finish my thoughts. I just copied the part of yours that is just a giant or clause

File: content/millennium/r4/request-and-response/communication.md (hunk 4 above)
Comment (sydneanschutz) [reply content/millennium/r4/request-and-response/communication.md:74]: Looking at appointment: http://fhir.cerner.com/millennium/r4/scheduling/appointment/#parameters
```
* The `received` parameter may be provided:
** once with the `ge` prefix.
** twice with the lower value prefixed by `ge` and the higher value prefixed by `le`.
```
shoutout to Sarah

Do we want to do the same thing here?

File: content/millennium/r4/other/binary.md (hunk 2 above)
Comment (sydneanschutz) [reply content/millennium/r4/other/binary.md]: https://github.com/cerner/fhir.cerner.com/pull/607/commits/73cd84784ce49a7b41ed138be9e5294f9ab8c264

**Review by VelocityKnight — APPROVED** [review]

File: content/millennium/r4/other/binary.md (hunk 1 above)
Comment (ebaranowski) [reply content/millennium/r4/other/binary.md]: +1

File: content/millennium/r4/other/binary.md (hunk 2 above)
Comment (ebaranowski) [reply content/millennium/r4/other/binary.md]: Im not in favor of the wording. Going to break down the parts I see:

The `Binary.read` scope and `Communication.read` scope or either the `DiagnosticReport.read` or the `DocumentReference.read` is required.

Binary.read` scope and `Communication.read` - this part seems independent of the rest of the sentence. Also seemingly makes it seem like Communication is required all the time.

or either - using this as a phrase is contradictory. Definition: used before the first of two (or occasionally more) alternatives that are being specified (the other being introduced by “or”).
"either I'll accompany you to your room, or I'll wait here"

Binary and Communication or either DiagnosticReport or DocumentReference = Binary and Communication or DiagnosticReport or DocumentReference
Which also means that the only required scope is Binary, and you just need one of the other scopes.

So at the end of this, Im still confused as to the direction this wording is going in.
I would think that just in it's current format you could do something like: 
```
The required scopes are the `Binary.read` scope, as well as one of the following scopes: `Communication.read`, `DiagnosticReport.read`, or `DocumentReference.read`.
```
and understand it better.

Comment (kathypickering) [reply content/millennium/r4/other/binary.md]: That's fine

Comment (sydneanschutz) [reply content/millennium/r4/other/binary.md]: https://github.com/cerner/fhir.cerner.com/pull/607/commits/1287fa77649531cdaf02fcc169b1a67612536562#diff-b9330978e62a15a61d6612ae8dc779e3b4680761c2dfc7e9b3778909d2d95cd8L45

File: content/millennium/r4/request-and-response/communication.md (hunk 4 above)
Comment (sydneanschutz) [reply content/millennium/r4/request-and-response/communication.md]: https://github.com/cerner/fhir.cerner.com/pull/607/commits/1287fa77649531cdaf02fcc169b1a67612536562#diff-cc6586eb639048bdea01208ce786612c64361622cd67bd53a5bf7253f27a840aL74

**Review by ebaranowski — APPROVED** [review]

File: content/millennium/r4/request-and-response/communication.md [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+
+<%= disclaimer %>
+
+### Errors
+
+The common [errors] and [OperationOutcomes] may be returned.
+
+[`reference`]: https://hl7.org/fhir/R4/search.html#reference
+[`token`]: https://hl7.org/fhir/R4/search.html#token
+[`date`]: http://hl7.org/fhir/R4/search.html#date
```
Comment (MadhurRaj) [inline content/millennium/r4/request-and-response/communication.md:245]: this and codeableconcept need to be https.

File: lib/resources/r4/communication.yaml [hunk 8]
```diff
@@ -0,0 +1,223 @@
+---
+name: Communication
+field_name_base_url: http://hl7.org/fhir/r4/communication-definitions.html#Communication
```
Comment (MadhurRaj) [inline lib/resources/r4/communication.yaml:3]: https.

File: lib/resources/r4/communication.yaml [hunk 9]
```diff
@@ -0,0 +1,223 @@
...
+    }
+  action:
+    - terminology
+    - create
+  binding:
+    description: The category of the communication.
+    terminology:
+      - display: Communication Category
+        system: http://terminology.hl7.org/CodeSystem/communication-category
+        info_link: http://hl7.org/fhir/R4/valueset-communication-category.html
```
Comment (MadhurRaj) [inline lib/resources/r4/communication.yaml:58]: https

File: lib/resources/r4/communication.yaml [hunk 10]
```diff
@@ -0,0 +1,223 @@
...
+      - display: SNOMED CT
+        system: http://snomed.info/sct
+        info_link: http://hl7.org/fhir/r4/snomedct.html
```
Comment (MadhurRaj) [inline lib/resources/r4/communication.yaml:61]: https

File: lib/resources/r4/communication_patch.yaml [hunk 11]
```diff
@@ -0,0 +1,33 @@
+---
+name: Communication
+field_name_base_url: http://hl7.org/fhir/R4/communication-definitions.html#Communication
```
Comment (MadhurRaj) [inline lib/resources/r4/communication_patch.yaml:3]: https

File: content/millennium/r4/request-and-response/communication.md (hunk 7 above)
Comment (sydneanschutz) [reply content/millennium/r4/request-and-response/communication.md]: https://github.com/cerner/fhir.cerner.com/pull/607/commits/396b90d0f3456da2649be53c3e9a900e7001e024#diff-cc6586eb639048bdea01208ce786612c64361622cd67bd53a5bf7253f27a840aL245

File: lib/resources/r4/communication.yaml (hunk 9 above)
Comment (sydneanschutz) [reply lib/resources/r4/communication.yaml]: https://github.com/cerner/fhir.cerner.com/pull/607/commits/396b90d0f3456da2649be53c3e9a900e7001e024#diff-246915dcfd8bc84060afc50e73fa123a82fdca704ca11eeffde54898c369b4cdL3

File: lib/resources/r4/communication.yaml (hunk 10 above)
Comment (sydneanschutz) [reply lib/resources/r4/communication.yaml]: https://github.com/cerner/fhir.cerner.com/pull/607/commits/396b90d0f3456da2649be53c3e9a900e7001e024#diff-246915dcfd8bc84060afc50e73fa123a82fdca704ca11eeffde54898c369b4cdL58

File: lib/resources/r4/communication.yaml (hunk 8 above)
Comment (sydneanschutz) [reply lib/resources/r4/communication.yaml]: https://github.com/cerner/fhir.cerner.com/pull/607/commits/396b90d0f3456da2649be53c3e9a900e7001e024#diff-246915dcfd8bc84060afc50e73fa123a82fdca704ca11eeffde54898c369b4cdL3

File: lib/resources/r4/communication_patch.yaml (hunk 11 above)
Comment (sydneanschutz) [reply lib/resources/r4/communication_patch.yaml]: https://github.com/cerner/fhir.cerner.com/pull/607/commits/396b90d0f3456da2649be53c3e9a900e7001e024#diff-a790f5f9d60e3713f13c08103ad6e6e260ef7a2eb11eecafd0db8f3451f1e3f7L3

**Review by MadhurRaj — APPROVED** [review]

**Review by kathypickering — APPROVED** [review]

**Comment by sujanb12** [comment]
Verified in live!
Screen Shot 2021-04-09 at 10 01 08 AM

**Pull request merged**